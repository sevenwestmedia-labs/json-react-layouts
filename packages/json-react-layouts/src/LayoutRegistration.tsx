import React from 'react'
import {
    ComponentInformation,
    ComponentRegistrationBuilderStart,
    ComponentRegistrationBuilder,
    ComponentRegistration,
    ComponentRegistrations,
} from './ComponentRegistrar'
import {
    CompositionInformation,
    CompositionRegistrations,
    CompositionRegistration,
} from './CompositionRegistrar'
import { RendererMiddleware, getRegistrationCreators, LayoutApi } from '.'
import { CompositionRenderer } from './renderers/composition-renderer'
import { CompositionsRenderer } from './renderers/compositions-renderer'
import { composeMiddleware } from './middlewares'
import { ComponentsRenderer } from './renderers/components-renderer'

interface RegisterComponentsStep<Services extends {}> {
    registerComponents<
        Components extends ComponentInformation<any, any>,
        ComponentMiddlewares extends {}
    >(
        registerCallback: (
            registrar: ComponentRegistrationBuilderStart<Services>,
        ) => ComponentRegistrationBuilder<Services, Components, ComponentMiddlewares>,
    ): RegisterCompositionsStep<Components, ComponentMiddlewares, Services>
}

export interface CompositionRegistrationBuilderStart<
    Components extends ComponentInformation<any, any>,
    ComponentMiddlewaresProps extends {},
    Services
> {
    registerComposition<TType extends string, TContentAreas extends string, TProps>(
        registration: CompositionRegistration<TType, TContentAreas, Services, TProps>,
    ): CompositionRegistrationBuilder<
        Components,
        ComponentMiddlewaresProps,
        CompositionInformation<TType, TContentAreas, TProps>,
        {},
        Services
    >
}

export interface CompositionRegistrationBuilder<
    Components extends ComponentInformation<any, any>,
    ComponentMiddlewaresProps extends {},
    Compositions extends CompositionInformation<any, any, any>,
    CompositionMiddlewaresProps extends {},
    Services
> {
    registerComposition<TType extends string, TContentAreas extends string, TProps>(
        registration: CompositionRegistration<TType, TContentAreas, Services, TProps>,
    ): CompositionRegistrationBuilder<
        Components,
        ComponentMiddlewaresProps,
        Compositions | CompositionInformation<TType, TContentAreas, TProps>,
        CompositionMiddlewaresProps,
        Services
    >

    registerMiddleware<TRegistrationMiddlewareProps extends {}>(
        compositionMiddleware: RendererMiddleware<Services, TRegistrationMiddlewareProps>,
    ): CompositionRegistrationBuilder<
        Components,
        ComponentMiddlewaresProps,
        Compositions,
        CompositionMiddlewaresProps & TRegistrationMiddlewareProps,
        Services
    >
}

interface RegisterCompositionsStep<
    Components extends ComponentInformation<any, any>,
    ComponentMiddlewaresProps extends {},
    Services extends {}
> {
    registerCompositions<
        Compositions extends CompositionInformation<any, any, any>,
        CompositionMiddlewaresProps extends {}
    >(
        registerCallback: (
            registrar: CompositionRegistrationBuilderStart<
                Components,
                ComponentMiddlewaresProps,
                Services
            >,
        ) => CompositionRegistrationBuilder<
            Components,
            ComponentMiddlewaresProps,
            Compositions,
            CompositionMiddlewaresProps,
            Services
        >,
    ): LayoutApi<
        Components,
        Compositions,
        ComponentMiddlewaresProps,
        CompositionMiddlewaresProps,
        Services
    >
}

export function LayoutRegistration<Services extends {}>(): RegisterComponentsStep<Services> {
    return {
        registerComponents<
            Components extends ComponentInformation<any, any>,
            ComponentMiddlewares extends {}
        >(
            registerCallback: (
                registrar: ComponentRegistrationBuilderStart<Services>,
            ) => ComponentRegistrationBuilder<Services, Components, ComponentMiddlewares>,
        ): RegisterCompositionsStep<Components, ComponentMiddlewares, Services> {
            // the internal collection of registered components
            const registeredComponents: {
                [key: string]: ComponentRegistration<any, any, Services>
            } = {}
            const componentMiddlewares: Array<RendererMiddleware<Services, any>> = []

            const builder: ComponentRegistrationBuilder<Services, any, any> = {
                registerComponent: registration => {
                    if (registeredComponents[registration.type]) {
                        throw new Error(`${registration.type} has already been registered`)
                    }

                    registeredComponents[registration.type] = registration
                    return builder
                },
                registerMiddleware: componentMiddleware => {
                    componentMiddlewares.push(componentMiddleware)
                    return builder
                },
            }

            registerCallback({
                registerComponent: registration => {
                    if (registeredComponents[registration.type]) {
                        throw new Error(`${registration.type} has already been registered`)
                    }

                    registeredComponents[registration.type] = registration
                    return builder
                },
            })

            // the following exposes registerCompositions
            return {
                registerCompositions<
                    Compositions extends CompositionInformation<any, any, any>,
                    CompositionMiddlewaresProps extends {}
                >(
                    registerCallback: (
                        registrar: CompositionRegistrationBuilderStart<
                            Components,
                            ComponentMiddlewares,
                            Services
                        >,
                    ) => CompositionRegistrationBuilder<
                        Components,
                        ComponentMiddlewares,
                        Compositions,
                        CompositionMiddlewaresProps,
                        Services
                    >,
                ): LayoutApi<
                    Components,
                    Compositions,
                    ComponentMiddlewares,
                    CompositionMiddlewaresProps,
                    Services
                > {
                    const registeredCompositions: {
                        [key: string]: CompositionRegistration<any, any, Services, any>
                    } = {}
                    const compositionMiddlewares: Array<RendererMiddleware<Services, any>> = []

                    const componentRegistrations: ComponentRegistrations = {
                        isRegistered(type: string) {
                            return registeredComponents[type] !== undefined
                        },
                        get(type: string) {
                            const foundComponent = registeredComponents[type]
                            if (!foundComponent && process.env.NODE_ENV !== 'production') {
                                // continue rendering in production only. otherwise throw, this is so the site does not crash
                                // empty area will be rendered instead
                                console.warn({ type }, `Component has not been registered`)
                            }
                            return foundComponent
                        },
                    }
                    const compositionRegistrations: CompositionRegistrations = {
                        get(type: string) {
                            const foundComposition = registeredCompositions[type]
                            if (!foundComposition && process.env.NODE_ENV !== 'production') {
                                // Warn a component is missing if not in production
                                console.warn({ type }, `Composition has not been registered`)
                            }
                            return foundComposition
                        },
                    }

                    const { createRegisterableComponent } = getRegistrationCreators<Services>()
                    const middlewares: {
                        component: RendererMiddleware<any, any>
                        composition: RendererMiddleware<any, any>
                    } = {
                        component: () => null,
                        composition: () => null,
                    }

                    const layout: LayoutApi<any, any, any, any, Services> = {
                        componentRegistrations,
                        compositionRegistrations,

                        component(component) {
                            return component as any
                        },
                        composition(composition) {
                            return composition as any
                        },
                        nestedComposition(composition) {
                            return {
                                type: 'nested-composition',
                                props: { composition: composition },
                            } as any
                        },

                        createRenderers({ services }) {
                            return {
                                renderComponents(...components) {
                                    return (
                                        <ComponentsRenderer
                                            layoutApi={layout}
                                            componentMiddleware={middlewares.component}
                                            componentRegistrations={componentRegistrations}
                                            services={services}
                                            components={components}
                                            componentRenderPath=""
                                            additionalComponentProps={{}}
                                        />
                                    )
                                },

                                renderCompositions(...compositions) {
                                    return (
                                        <CompositionsRenderer
                                            layoutApi={layout}
                                            componentMiddleware={middlewares.component}
                                            compositionMiddleware={middlewares.composition}
                                            componentRegistrations={componentRegistrations}
                                            compositionRegistrations={compositionRegistrations}
                                            services={services}
                                            compositions={compositions}
                                        />
                                    )
                                },
                            }
                        },
                    }

                    builder.registerComponent(
                        createRegisterableComponent(
                            'nested-composition',
                            (
                                {
                                    composition,
                                    componentRenderPath,
                                    ...additionalComponentProps
                                }: NestedCompositionProps<any>,
                                services,
                            ) => {
                                if (composition.type === 'nested-composition') {
                                    throw new Error(
                                        "nested-composition is registered as a component, it is not allowed within a nested composition as it's not a composition, please check the route",
                                    )
                                }

                                return (
                                    <CompositionRenderer
                                        composition={composition}
                                        componentRenderPath={`${componentRenderPath}nested:${composition.type}`}
                                        services={services}
                                        layoutApi={layout}
                                        componentRegistrations={componentRegistrations}
                                        compositionRegistrations={compositionRegistrations}
                                        componentMiddleware={middlewares.component}
                                        compositionMiddleware={middlewares.composition}
                                        additionalComponentProps={additionalComponentProps}
                                    />
                                )
                            },
                        ),
                    )

                    const compositionBuilder: CompositionRegistrationBuilder<
                        any,
                        any,
                        any,
                        any,
                        Services
                    > = {
                        registerComposition(registration) {
                            if (registeredCompositions[registration.type]) {
                                throw new Error(`${registration.type} has already been registered`)
                            }

                            registeredCompositions[registration.type] = registration
                            return compositionBuilder
                        },
                        registerMiddleware(compositionMiddleware) {
                            compositionMiddlewares.push(compositionMiddleware)
                            return compositionBuilder
                        },
                    }

                    registerCallback(compositionBuilder)

                    middlewares.component = composeMiddleware(componentMiddlewares)
                    middlewares.composition = composeMiddleware(compositionMiddlewares)

                    return layout
                },
            }
        },
    }
}

export interface NestedCompositionProps<
    Compositions extends CompositionInformation<any, any, any>
> {
    composition: Compositions
    componentRenderPath?: string
}

export type ComponentCheckedMessage = 'wrap in layout.component()'
export type CompositionCheckedMessage = 'wrap in layout.composition()'
