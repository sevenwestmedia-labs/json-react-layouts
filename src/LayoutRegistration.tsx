import React from 'react'
import { ComponentRegistrar, ComponentInformation } from './ComponentRegistrar'
import {
    CompositionRegistrar,
    CompositionInformation,
    NestedCompositionProps,
} from './CompositionRegistrar'
import { LayoutApi } from './LayoutApi'
import { getRegistrationCreators } from './get-registration-creators'
import { Logger, noopLogger } from 'typescript-log'
import { ComponentRenderer } from './ComponentRenderer'

export class LayoutRegistration<Services = {}> {
    constructor(private logger: Logger = noopLogger()) {}

    registerComponents<
        Components extends ComponentInformation<any, any> = never,
        ComponentMiddlewares extends object = {}
    >(
        registerCallback: (
            registrar: Pick<
                ComponentRegistrar<Services, never, {}>,
                'registerComponent' | 'registerMiddleware'
            >,
        ) => Pick<
            ComponentRegistrar<Services, Components, ComponentMiddlewares>,
            'registerComponent' | 'registerMiddleware'
        >,
    ): LayoutCompositionRegistration<Components, ComponentMiddlewares, Services> {
        const regisrar = new ComponentRegistrar<Services, never, {}>(this.logger)

        const unmaskedComponentRegistrar = registerCallback(regisrar) as ComponentRegistrar<
            Services,
            Components,
            ComponentMiddlewares
        >

        return new LayoutCompositionRegistration(unmaskedComponentRegistrar)
    }
}

type NestedComponentInfo = ComponentInformation<'nested-composition', NestedCompositionProps>

export class LayoutCompositionRegistration<
    Components extends ComponentInformation<any>,
    ComponentMiddlewaresProps extends {},
    Services
> {
    constructor(
        private componentRegistrar: ComponentRegistrar<
            Services,
            Components,
            ComponentMiddlewaresProps
        >,
    ) {}

    /** Returns a components renderer component which can render a list of components */
    createComponentsRenderer(): React.FC<{
        components: Array<Components & ComponentMiddlewaresProps>
        services: Services
    }> {
        const layoutApi = this.registerCompositions(registrar => registrar) as any

        interface ComponentsRendererProps {
            components: Array<Components & ComponentMiddlewaresProps>
            services: Services
        }

        const ComponentsRenderer: React.FC<ComponentsRendererProps> = ({
            components,
            services,
        }) => {
            return (
                <React.Fragment>
                    {components.map((item, index) => {
                        const { type, props: componentProps, ...middlewareProps } = item
                        return (
                            <ComponentRenderer
                                key={`${item.type}-${index}`}
                                type={type}
                                layoutApi={layoutApi}
                                componentRegistrar={this.componentRegistrar}
                                componentProps={{
                                    ...componentProps,
                                    componentType: item.type,
                                    componentRenderPath: `[${index}]`,
                                }}
                                middlewareProps={middlewareProps}
                                services={services}
                                renderComponentMiddleware={
                                    this.componentRegistrar.componentMiddleware
                                }
                            />
                        )
                    })}
                </React.Fragment>
            )
        }
        ComponentsRenderer.displayName = 'ComponentsRenderer'

        return ComponentsRenderer
    }

    registerCompositions<Compositions extends CompositionInformation<any, any, any>>(
        registerCallback: (
            registrar: Pick<
                CompositionRegistrar<
                    Components | ComponentInformation<'nested-composition', NestedCompositionProps>,
                    Services,
                    ComponentMiddlewaresProps
                >,
                'registerComposition'
            >,
        ) => Pick<
            CompositionRegistrar<
                Components | ComponentInformation<'nested-composition', NestedCompositionProps>,
                Services,
                ComponentMiddlewaresProps,
                Exclude<Compositions, never>
            >,
            'registerComposition'
        >,
    ): LayoutApi<
        Components | NestedComponentInfo,
        Exclude<Compositions, never>,
        Services,
        ComponentMiddlewaresProps
    > {
        const { createRegisterableComponent } = getRegistrationCreators<Services>()

        const registrarWithNestedComposition = this.componentRegistrar.registerComponent(
            createRegisterableComponent(
                'nested-composition',
                (props: NestedCompositionProps, services) => {
                    if (props.composition.type === 'nested-composition') {
                        throw new Error(
                            "nested-composition is registered as a component, it is not allowed within a nested composition as it's not a composition, please check the route",
                        )
                    }

                    return (
                        <CompositionRenderer
                            compositionInformation={props.composition}
                            componentRenderPath={`${props.componentRenderPath}/nested:${props.composition.type}`}
                            services={services}
                            layoutApi={layout}
                        />
                    )
                },
            ),
        )
        const registrar = new CompositionRegistrar<
            Components | NestedComponentInfo,
            Services,
            ComponentMiddlewaresProps
        >(registrarWithNestedComposition as any)

        // We use the mast to limit the API the end user sees, cast back to full compositionRegistrar
        const unmaskedCompositionRegistrar = registerCallback(registrar) as CompositionRegistrar<
            Components | ComponentInformation<'nested-composition', NestedCompositionProps>,
            Services,
            ComponentMiddlewaresProps,
            Exclude<Compositions, never>
        >

        const layout = new LayoutApi(unmaskedCompositionRegistrar)
        const CompositionRenderer = layout.CompositionRenderer
        return layout
    }
}
