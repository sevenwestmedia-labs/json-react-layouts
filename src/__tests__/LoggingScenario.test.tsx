import React from 'react'
import { configure, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { Logger } from 'typescript-log'
import {
    testComponentRegistration,
    testComponent2Registration,
    testCompositionRegistration,
} from './testComponents'
import { LayoutRegistration } from '../LayoutRegistration'

configure({ adapter: new Adapter() })

it('can render a composition with a content area through the registrar', () => {
    const logMessages: any[] = []
    const addToLog = (...args: any[]) =>
        logMessages.push(
            args.reduce<string>(
                (acc, val) => (acc ? acc + ', ' + JSON.stringify(val) : JSON.stringify(val)),
                '',
            ),
        )
    const logger: Logger = {
        child: () => logger,
        debug: addToLog,
        error: addToLog,
        fatal: addToLog,
        info: addToLog,
        trace: addToLog,
        warn: addToLog,
    }

    const layout = new LayoutRegistration(logger)
        .registerComponents(registrar =>
            registrar
                .registerComponent(testComponentRegistration)
                .registerComponent(testComponent2Registration),
        )
        .registerCompositions(registrar =>
            registrar.registerComposition(testCompositionRegistration),
        )

    mount(
        <layout.CompositionsRenderer
            services={{}}
            compositions={[
                {
                    type: 'test-composition',
                    props: {},
                    contentAreas: {
                        main: [
                            { type: 'test', props: {} },
                            layout.nestedComposition({
                                type: 'test-composition',
                                props: {},
                                contentAreas: { main: [{ type: 'test', props: {} }] },
                            }),
                        ],
                    },
                },
            ]}
        />,
    )

    expect(logMessages).toMatchSnapshot()
})
