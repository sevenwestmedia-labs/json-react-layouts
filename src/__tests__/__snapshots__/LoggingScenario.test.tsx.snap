// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can render a composition with a content area through the registrar 1`] = `
Array [
  "{\\"compositions\\":[{\\"type\\":\\"test-composition\\"}]}, \\"Rendering compositions\\"",
  "{\\"componentRenderPath\\":\\"homepage[0-test-composition]\\",\\"type\\":\\"test-composition\\"}, \\"Rendering composition\\"",
  "{\\"componentRenderPath\\":\\"homepage[0-test-composition]/main\\",\\"components\\":[{\\"type\\":\\"test\\"},{\\"type\\":\\"nested-composition\\"}]}, \\"Rendering content area\\"",
  "{\\"componentRenderPath\\":\\"homepage[0-test-composition]/main[1]/nested:test-composition\\",\\"type\\":\\"test-composition\\"}, \\"Rendering composition\\"",
  "{\\"componentRenderPath\\":\\"homepage[0-test-composition]/main[1]/nested:test-composition/main\\",\\"components\\":[{\\"type\\":\\"test\\"}]}, \\"Rendering content area\\"",
]
`;
